[inst-simplify] Update for OSSA

gottesmm opened the pull request:
I am splitting this out and fixing some issues at atrick's request.

meg-gupta commented on lib/SIL/IR/SILInstruction.cpp:
```diff
@@ -1029,6 +1029,31 @@ SILInstruction::MemoryBehavior SILInstruction::getMemoryBehavior() const {
+      return MemoryBehavior::MayHaveSideEffects;
+    }
+    llvm_unreachable("Covered switch isn't covered?!");
+  }
```
We may also need to update  `MemoryBehaviorVisitor::visitLoadInst` and  `MemoryBehaviorVisitor::visitStoreInst` for [take] and [assign] ?

gottesmm replied:
Yes. That is what I brought up here: https://github.com/apple/swift/pull/34543.

atrick commented on include/swift/SILOptimizer/Utils/ValueLifetime.h:
```diff
@@ -54,24 +55,47 @@ class ValueLifetimeAnalysis {
   template <typename RangeTy>
-  ValueLifetimeAnalysis(decltype(defValue) def, const RangeTy &userRange)
-      : defValue(def), userSet(userRange.begin(), userRange.end()) {
+  ValueLifetimeAnalysis(SILArgument *def, const RangeTy &useRange)
```
Why is this renamed? `useRange` should have Operand types, but this still expects instruction pointers, or is there supposed to be a call to `getUser` here?

Also, why do we require a list of use Operands if we're only storing the set of user instructions?

atrick commented on lib/SIL/Utils/BasicBlockUtils.cpp:
```diff
@@ -372,3 +372,146 @@ void DeadEndBlocks::compute() {
+  }
+}
+
+bool swift::completePostDominatingSet(
```
Is there a reason this API doesn't return a move-only result
```
struct Result {
  bool hasLoop;
  SmallPtrSet<SILBasicBlock *, C>
};
```

gottesmm replied:
From our conversation: I'll add a comment explaining what is the input/output of the utility.

atrick commented on lib/SIL/Utils/BasicBlockUtils.cpp:
```diff
@@ -372,3 +372,146 @@ void DeadEndBlocks::compute() {
+}
+
+bool swift::completePostDominatingSet(
+    SILBasicBlock *definingBlock, SILBasicBlock *reachableBlock,
```
It seems odd for a utility to expect a single `reachableBlock`. In most cases don't we have a set of reachable blocks?

atrick commented on lib/SIL/Utils/BasicBlockUtils.cpp:
```diff
@@ -372,3 +372,146 @@ void DeadEndBlocks::compute() {
+
+} // anonymous namespace
+
+void PostDominanceCompletionSetComputer::initializeDataflow() {
```
This algorithm does not have a separate "initialization" step, which implies something separate from the worklist processing. This function should just be a `pushPredecessors` utility called by the main `performDataflow`, once before entering the loop and once inside. It's actually alarming that it doesn't do the same thing, e.g. mark the predecessor visited.

atrick commented on lib/SIL/Utils/BasicBlockUtils.cpp:
```diff
@@ -372,3 +372,146 @@ void DeadEndBlocks::compute() {
+  }
+}
+
+void PostDominanceCompletionSetComputer::performDataflow() {
```
I guess you could call the "data flow" but it's a little confusing because all it's doing is marking blocks visited. It just finds reachable blocks.

atrick reviewed (commented):
Reviewed the first half, not including SimplifyInstruction or SILCombine

atrick commented on lib/SILOptimizer/Analysis/SimplifyInstruction.cpp:
```diff
@@ -727,45 +746,556 @@ SILValue swift::simplifyInstruction(SILInstruction *I) {
+  }
+}
+
+static ValueOwnershipKind findCompatibleNonNoneOwnershipKind(OperandOwnershipKindMap map) {
```
I'm worried to see new code dealing with `OwnershipKindMaps` instead of a simpler lifetime constraint model. The kind maps severely obscure the ownership model and I'm afraid it will be more difficult to eliminate the kind maps later if they keep getting propagated through APIs instead of being encapsulated behind a simple interface for "ownership constraints" ('owned', 'guaranteed', 'any') plus "lifetime constraints" ('consuming', 'immediate', 'scoped', 'propagating/escaping').

In this case, the entire `findCompatibleNonNoneOwnershipKind` helper should just be a trivial lattice 'join' over the new ownership kind and the operand's ownership constraint.

atrick commented on lib/SILOptimizer/Analysis/SimplifyInstruction.cpp:
```diff
@@ -727,45 +746,556 @@ SILValue swift::simplifyInstruction(SILInstruction *I) {
+  assert(onlyAcceptNone.empty() && convertToOwned.empty() && convertToGuaranteed.empty() && convertToUnowned.empty());
+
+  // Then handle the non-conversion uses.
+  switch (newKind) {
```
I don't know why you need to separately switch over the operand ownership constraint above and switch over the new kind here. The new operand kind should be determined directly by joining the new value kind with the operand constraint. Then you switch over OwnershipKindPair reusing the same code in `::perform` but without any of the ".none" cases.

atrick reviewed (dismissed):
I don't think you should handle .none as an ownership kind with special treatment. Converting to .none can be trivially handled before doing anything fancy. Converting from .none involves deciding the ownership of each use, which should be a trivial lattice join between the new kind and the operand's constraint. Then partitioning those uses uses into two (or three) separate kinds, and for each category reusing the same code "as-if" the original .none value were either owned or guaranteed.

gottesmm commented:
Going to rebase this.

gottesmm committed:
[semantic-arc] Change SemanticARCOptVisitor to be a SILValueVisitor instead of a SILInstructionVisitor.

Our worklist already uses values, so it makes sense to use a SILValueVisitor. It
will also let me throw in a few ARC dead argument elimination optimizations so
clean up a pattern from my ARC RAUW thing. Once we have simplify-cfg it will not
be needed, but I want to know I am good before I flip the switch.

gottesmm committed:
[ownership] Commit a generic replaceAllUsesAndEraseFixingOwnership api and enable SimplifyInstruction on OSSA.

This is a generic API that when ownership is enabled allows one to replace all
uses of a value with a value with a differing ownership by transforming/lifetime
extending as appropriate.

This API supports all pairings of ownership /except/ replacing a value with
OwnershipKind::None with a value without OwnershipKind::None. This is a more
complex optimization that we do not support today. As a result, we include on
our state struct a helper routine that callers can use to know if the two values
that they want to process can be handled by the algorithm.

My moticiation is to use this to to update InstSimplify and SILCombiner in a
less bug prone way rather than just turn stuff off.

Noting that this transformation inserts ownership instructions, I have made sure
to test this API in two ways:

1. With Mandatory Combiner alone (to make sure it works period).

2. With Mandatory Combiner + Semantic ARC Opts to make sure that we can
   eliminate the extra ownership instructions it inserts.

As one can see from the tests, the optimizer today is able to handle all of
these transforms except one conditional case where I need to eliminate a dead
phi arg. I have a separate branch that hits that today but I have exposed unsafe
behavior in ClosureLifetimeFixup that I need to fix first before I can land
that. I don't want that to stop this PR since I think the current low level ARC
optimizer may be able to help me here since this is a simple transform it does
all of the time.

gottesmm committed:
[semantic-arc] Eliminate unneeded unchecked_ownership_conversion from owned/guaranteed -> unowned.

Specifically, we check that all of the unowned's value uses are within the
lifetime of the unchecked_ownership_conversion's operand. If so, we eliminate
it.

gottesmm commented:
I upstreamed and slimmed this down. I did it the way we talked about offline.

gottesmm commented:
I think I also redid everything that were a problem.

gottesmm commented on include/swift/SILOptimizer/Utils/CanonicalizeInstruction.h:
```diff
@@ -63,7 +63,8 @@ struct CanonicalizeInstruction {
   /// erases \p inst, then it may be an invalid pointer upon return.
-  SILBasicBlock::iterator canonicalize(SILInstruction *inst);
+  SILBasicBlock::iterator canonicalize(SILInstruction *inst,
+                                       DeadEndBlocks &deadEndBlocks);
```
I think the issue was that I don't have access to the function in Canonicalization. So I need to pass it in.

gottesmm replied:
Ok. Let me look.

gottesmm replied:
Fixed it locally.

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci benchmark

gottesmm commented:
@swift-ci test source compatibility

gottesmm commented:
preset=buildbot,tools=RA,stdlib=RD,test=non_executable
@swift-ci Please test with preset macOS

gottesmm commented:
I am going to have to fight a little bit with MSVC looks like.

gottesmm commented:
I am really curious what the benchmark results turn out. I diffed the text and there isn't any difference at -O anyways.

gottesmm commented:
If all of the tests pass, I am going to fix the windows issue and then smoke test this in.

atrick commented on include/swift/SILOptimizer/Utils/OwnershipUtils.h:
```diff
@@ -0,0 +1,51 @@
+//===--- OwnershipUtils.h -------------------------------------------------===//
```
Duplicate filename, easily distinguished as `OwnershipOptUtils.h`

gottesmm replied:
Good idea! Fixing!

atrick commented on lib/SILOptimizer/Utils/OwnershipUtils.cpp:
```diff
@@ -0,0 +1,804 @@
+//===----------------------------------------------------------------------===//
+
+static SILBasicBlock::iterator
+replaceAllUsesAndEraseInner(SingleValueInstruction *svi, SILValue newValue,
```
Can there be just one copy of `replaceAllUsesAndEraseInner`?

atrick commented on lib/SILOptimizer/Utils/OwnershipUtils.cpp:
```diff
@@ -0,0 +1,804 @@
+  if (oldValue->getModule().getStage() == SILStage::Raw)
+    return false;
+
+  // In constrast, if our old ownership kind is ValueOwnershipKind::None and
```
These comments are hard to parse because the description is the inverse of the condition being commented. You could either flip the condition in the code or delete the first sentence in this comment.

gottesmm replied:
I fixed it.

atrick reviewed (commented):
Looks good overall. I'm still reviewing the RAUW utility and lifetime extender.

gottesmm commented on test/SILOptimizer/semantic-arc-opt-unchecked-ownership-conversion.sil:
```diff
@@ -0,0 +1,248 @@
+// CHECK-LABEL: sil [ossa] @owned_to_unowned_negative_1 : $@convention(thin) (@guaranteed Klass) -> () {
+// CHECK: unchecked_ownership_conversion
+// CHECK: } // end sil function 'owned_to_unowned_negative_1'
+sil [ossa] @owned_to_unowned_negative_1 : $@convention(thin) (@guaranteed Klass) -> () {
+bb0(%0 : @guaranteed $Klass):
+  %1 = copy_value %0 : $Klass
+  %2 = unchecked_ownership_conversion %1 : $Klass, @owned to @unowned
+  destroy_value %1 : $Klass
```
Nothing. Unowned means that you have to retain it before you use it as an owned or guaranteed value. Otherwise, you can just pass it off.

gottesmm commented on include/swift/SILOptimizer/Utils/OwnershipUtils.h:
```diff
@@ -0,0 +1,51 @@
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
```
It is standard style according to LLVM style. I am just going with that.

gottesmm commented on lib/SIL/Utils/OwnershipUtils.cpp:
```diff
@@ -221,9 +221,17 @@ bool BorrowingOperand::visitLocalEndScopeUses(
   }
+  }
+
+  llvm_unreachable("Covered switch isn't covered");
```
The reason to have this is so that if someone is lazy, ignores the compiler, we get an assert. I don't think it hurts and also shows that the code writer intended this code to always be covered. So I think it is reasonable.

gottesmm commented on lib/SILOptimizer/Mandatory/MandatoryCombine.cpp:
```diff
@@ -93,9 +95,9 @@ class MandatoryCombineCanonicalize final : CanonicalizeInstruction {
   }
 
-  bool tryCanonicalize(SILInstruction *inst) {
+  bool tryCanonicalize(SILInstruction *inst, DeadEndBlocks &deadEndBlocks) {
```
I fixed it.

gottesmm commented on lib/SILOptimizer/SemanticARC/OwnershipConversionElimination.cpp:
```diff
@@ -0,0 +1,67 @@
+bool SemanticARCOptVisitor::visitUncheckedOwnershipConversionInst(
+    UncheckedOwnershipConversionInst *uoci) {
+  // Return false if we are supposed to only be running guaranteed opts.
+  if (ctx.onlyGuaranteedOpts)
```
Why don't we talk about this offline. I'll do a follow up patch to rename to whatever makes sense.

gottesmm commented on lib/SILOptimizer/Utils/OwnershipUtils.cpp:
```diff
@@ -0,0 +1,804 @@
+//===--- OwnershipUtils.cpp -----------------------------------------------===//
```
Fixed.

gottesmm commented on lib/SILOptimizer/SemanticARC/OwnershipConversionElimination.cpp:
```diff
@@ -0,0 +1,67 @@
+  if (ctx.onlyGuaranteedOpts)
+    return false;
+
+  // Then check if we are running tests and shouldn't perform this optimiuzation
```
Fixed.

gottesmm commented on lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp:
```diff
@@ -79,12 +79,8 @@ killInstAndIncidentalUses(SingleValueInstruction *inst,
-    return None;
-
+simplifyAndReplace(SILInstruction *inst, CanonicalizeInstruction &pass,
+                   DeadEndBlocks &deadEndBlocks) {
```
Fixed.

gottesmm commented:
@swift-ci smoke test

gottesmm commented:
Done... sooo happy!
